rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,228 @@
+defined as a configuration value (`xpack.encrypted_saved_objects.encryptionKey`) via `kibana.yml`, but it's **highly 
+recommended** to define this key in the [Kibana Keystore](https://www.elastic.co/guide/en/kibana/current/secure-settings.html)
+instead. The master key should be cryptographically safe and be equal or greater than 32 bytes.
+
+To prevent certain vectors of attacks where raw content of encrypted attributes of one saved object is copied to another
+saved object that would unconsciously allow to decrypt content that was not supposed to be decrypted we rely on Additional
+authenticated data (AAD) during encryption and decryption. AAD consists of the following components:
+
+* Saved object ID
+* Saved object type
+* Saved object attributes
+ 
+AAD does not include encrypted attributes themselves and attributes defined in optional `attributesToExcludeFromAAD` 
+parameter provided during saved object type registration with `encrypted_saved_objects` plugin.
+
+## Audit
+
+Encrypted attributes will most likely contain a sensitive information and any attempt to access these should be properly
+logged to allow any further audit procedures. The following events will be logged with Kibana audit log functionality:
+
+* Successful attempt to encrypt attributes (incl. saved object ID, type and attributes names)
+* Failed attempt to encrypt attribute (incl. saved object ID, type and attribute name)
+* Successful attempt to decrypt attributes (incl. saved object ID, type and attributes names)
+* Failed attempt to decrypt attribute (incl. saved object ID, type and attribute name)
+
+In addition to audit log events we'll issue ordinary log events for any attempts to save, update or decrypt saved objects
+with missing attributes that were supposed to be encrypted/decrypted based on the registration parameters. 
+
+# Benefits
+
+* None of the registered types will expose their encrypted details. The saved
+objects with their unencrypted attributes could still be obtained and searched
+on. The wrapper will follow all the security and spaces filtering of saved
+objects so that only users with appropriate permissions will be able to obtain
+the scrubbed objects or _save_ objects with encrypted attributes.
+
+* No explicit access to a method that takes in an encrypted string exists. And no
+access to the private key used is exposed through the plugin. If the type was not
+registered no decryption is possible. No need to handle the saved object with
+the encrypted attributes reducing the risk of accidentally returning it in a
+handler.
+
+# Drawbacks
+
+* Possibly have a performance impact on Saved Objects API operations that require encryption/decryption
+* Will require non trivial tests to test functionality along with spaces and security
+* The attributes that are encrypted have to be defined and if they change they need to be migrated
```

Comment:
Another drawback is that changing an encryption key will permanently break the existing encrypted attributes, effectively losing them.

Reply:
Hmm, yeah, good point, I'll add here. I considered that as a limitation ("out of scope") of the RFC as I haven't heard any request for it to be in the initial version. API itself doesn't have such a flaw/drawback, it allows us to implement rotation mechanism on top of it once we need it. 

Key rotation deserves its own full blown RFC, e.g.:
```yaml
xpack.eso.encryptionKey:
  primary: 82Ffupq1uN/rlLIFHyUM7GkJcAAtCCPX #used to encrypt new data
  secondary: # used _only_ to decrypt old data
    - PPPtGN4Fh/aEQlqAC83BbeN9iQzzcMlO
    - 7Xso1va1kIvFW6xv8C9LH8crf2EVDVms
```
plus a privileged API endpoint for an emergency rotation that would re-encrypt existing data
```HTTP
POST /api/v1/eso/rotate/[optional saved object type]
```

Reply:
Given that we don't currently support seamlessly rotating the encryption key for Reporting or Security's usage, I agree with considering this requirement out of scope. 

The only situations that I'm aware of where we presently recommend changing the encryption key are when it's been compromised/leaked, and in those situations, we'd need to change the encryption key and consider all data using the old encryption key invalid because it could have been manipulated using the compromised encryption key.